### PR TITLE
[DPE-4942] - add mongos charm basics pt2 - add more suggestions from Carl

### PR DESCRIPTION
**Pull request by MiaAltieri**
## Issue
Our initial base of our charm was missing suggestions from Carl

## Solution
Add those suggestions

## Check libs 
This will keep failing as `charms.mongos.v0.mongos_client_interface` is not up to date, but it will be fixed as part of [DPE-5005](https://warthogs.atlassian.net/browse/DPE-5005)

[DPE-5005]: https://warthogs.atlassian.net/browse/DPE-5005?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tox.ini
Comment (Gu1nness) [inline tox.ini:1]: This should be updated to run ruff instead of isort + black + pflake8 in the `lint` environment

File: charmcraft.yaml [hunk 1]
```diff
@@ -1,20 +1,27 @@
...
+      craftctl default
+      if [[ ! -f requirements.txt ]]
+      then
+          echo 'ERROR: Use "tox run -e build-dev" instead of calling "charmcraft pack" directly' >&2
+          exit 1
+      fi
+    override-build: |
+      rustup default stable
+      craftctl default
+    charm-strict-dependencies: true
```
Comment (Gu1nness) [inline charmcraft.yaml:21]: This raises:
```
 Bad charmcraft.yaml content:
- 'charm-strict-dependencies' requires at least one requirements file. in field 'parts.charm.charm-strict-dependencies'
```
in the CI

**Review by Gu1nness — CHANGES_REQUESTED** [review]

**Review by Gu1nness — APPROVED** [review]

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -10,11 +10,11 @@ jobs:
     uses: ./.github/workflows/ci.yaml
     secrets: inherit
     permissions:
-      contents: write  # Needed for Allure Report beta
+      contents: write # Needed for Allure Report beta
 
   build:
     name: Build charm
-    uses: canonical/data-platform-workflows/.github/workflows/build_charm_without_cache.yaml@v16.7.0
+    uses: canonical/data-platform-workflows/.github/workflows/build_charm.yaml@v16.3.0
```
Comment (carlcsaposs-canonical) [inline .github/workflows/release.yaml]: please use same version of dpw for all workflows (ideally use latest version)

compatibility not guaranteed between workflows/python packages with different versions

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -50,27 +55,23 @@ jobs:
...
-        path:
-          - .
+    name: Build charm
+    uses: canonical/data-platform-workflows/.github/workflows/build_charm.yaml@v18.0.0
     with:
       cache: true
-      path-to-charm-directory: ${{ matrix.path }}
+
   integration-test:
-    name: Integration test charm | 3.4.4
```
Comment (Mehdi-Bendriss) [inline .github/workflows/ci.yaml:63]: Could we keep this? It is helpful to quickly know what juju version we're relying on, so we know when to bump

File: pyproject.toml [hunk 4]
```diff
@@ -2,29 +2,17 @@
 # See LICENSE file for licensing details.
 
 [tool.poetry]
-name = "mongos-k8s"
```
Comment (Mehdi-Bendriss) [inline pyproject.toml:5]: why delete all this part?

File: tox.ini [hunk 5]
```diff
@@ -41,8 +41,8 @@ commands_pre =
     poetry install --only format
 commands =
     poetry lock --no-update
-    poetry run isort {[vars]all_path}
-    poetry run black {[vars]all_path}
+    poetry run ruff check  --fix {[vars]all_path}
```
Comment (Mehdi-Bendriss) [inline tox.ini:44]: nice

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thanks Mia - looks good, I left a couple of minor comments

File: pyproject.toml (hunk 4 above)
Comment (MiaAltieri) [reply pyproject.toml:5]: These fields are used for when we are in package mode. But a charm is not a package so these fields are not necessary.

This was a request from Carl as to follow best practices

**Comment by Gu1nness** [comment]
> This will keep failing as charms.mongos.v0.mongos_client_interface is not up to date, but it will be fixed as part of [DPE-5003](https://warthogs.atlassian.net/browse/DPE-5003)

this is not DPE-5003 which is about upgrades on MongoDB VM

[DPE-5003]: https://warthogs.atlassian.net/browse/DPE-5003?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

File: pyproject.toml (hunk 4 above)
Comment (MiaAltieri) [reply pyproject.toml:5]: These fields are used for when we are in package mode. But a charm is not a package so these fields are not necessary.

This was a request from Carl as to follow best practices

**Review by Mehdi-Bendriss — APPROVED** [review]

**Pull request merged**